### DISCUSSION TITLE
Best way to support react-native?

### DISCUSSION THREAD

**Discussion by bi3ri** [rehypejs/rehype#162] [Q&A] [answered]
Hello and happy new year!

As the title suggests, I want to use remark / rehype to render markdown in a React Native application. I've already read about some other attempts, but the problem so far was how rehype handled text elements, [so they can be transformed into components](https://github.com/rehypejs/rehype-react/issues/17).

Now, my question is: What would be the best/easiest solution for this?

From my point of view, I see two options:
a) Write something like remark-react-native and skip rehype altogether. Something similar was done with remark-react, but it was deprecated because it was less pluggable and missing the rich plugin ecosystem of rehype.
b) Write a plugin that can transform text into elements.

The unified ecosystem can be a bit intimidating with all its plugins, different ASTs, parsers, and processors, so I am happy for any suggestion and wink in the right direction! :)

Thanks in advance!
Johannes

**Comment by ChristianMurphy** [comment, marked as answer]
Welcome @bi3ri! 👋 

As with so many things in Engineering "best" and "easiest" are very relative to what your goal is. 😉 🙂 

For easiest, as in least code to maintain, you could leverage the existing `rehype` ecosystem, and add an HTML to react-native translator on the tail end.
There are some existing solutions like https://meliorence.github.io/react-native-render-html/ which can take HTML text and render it in `react-native`.

If you want to avoid serializing to text before rendering.
I'd recommend leaning on https://github.com/rehypejs/rehype-react (a thing wrapper around https://github.com/syntax-tree/hast-util-to-jsx-runtime).
Rather than focusing on how to render, focus on how to adjust/rework the HTML AST to make it compatible with the `react-native` JSX runtime.
There are guides at https://unifiedjs.com/learn/ which introduce how to work with syntax trees.

**Reply by bi3ri** [reply]
Thanks a lot for your answer @ChristianMurphy! :pray: 
~I think for now I'll go with `react-native-render-html` and switch later to a `rehype-react` plugin.~

Edit:
Unfortunately to make `react-native-render-html` work for my case I need to many workarounds, that it would be worth it.

I am trying to adapt `rehype-react` for `react-native`, but so far without success. It seems that I can only redefine `tag-names` but not define my own components.

```javascript
const components = {
  // this works
  h1: "h2",
  // this does not work
  h2: (props) => {
    const { node } = props
    return <Text style={{ fontSize: 99 }}>{node.value}</Text>
  },
}
``` 

Am I doing sth worng in the `h2` case or is this not supported?
Would it be maybe better to write my own processor for `react-native`?

**Reply by bi3ri** [reply]
@ChristianMurphy: Could you give me a hint again for my question above? :)

**Reply by wooorm** [reply]
Add some console.logs in your functions to see what your functions are being passed and what you are doing wrong.
Make a reproduction of your case otherwise. See https://github.com/rehypejs/.github/blob/main/support.md for more info

**Reply by ChristianMurphy** [reply]
> Unfortunately to make `react-native-render-html` work for my case I need to many workarounds, that it would be worth it.

It may be worth discussing with the `react-native-render-html` team to make the renderer more robust.

> I am trying to adapt rehype-react for react-native, but so far without success.

As @wooorm notes above more detail is needed, help us help you https://github.com/rehypejs/.github/blob/main/support.md

> It seems that I can only redefine tag-names but not define my own components.

Are you seeing a typescript error or a runtime error?
Please share a reproducible example https://stackoverflow.com/help/minimal-reproducible-example

> Would it be maybe better to write my own processor for react-native?

You probably will, both markdown and HTML allow text in places that react-native does not.
You'll need to restructure text to ensure it is wrapped in a `<Text>` component.

Related previous discussions:
* https://github.com/orgs/remarkjs/discussions/1002
* https://github.com/orgs/remarkjs/discussions/1183

**Reply by ChristianMurphy** [reply]
It's also worth noting web views also exist I'm React native and can display browser rendered HTML snippets in native content https://github.com/react-native-webview/react-native-webview

**Reply by bi3ri** [reply]
>  Add some console.logs in your functions to see what your functions are being passed and what you are doing wrong.
>  Make a reproduction of your case otherwise. See https://github.com/rehypejs/.github/blob/main/support.md for more info

@wooorm: I created a reproducible example [here](https://github.com/bi3ri/rehype-react-native-demo). 

--- 

> It may be worth discussing with the react-native-render-html team to make the renderer more robust.

The project is unfortunately no longer maintained / there is no response to open issues. 

> As @wooorm notes above more detail is needed, help us help you https://github.com/rehypejs/.github/blob/main/support.md

I added a reproducible example above. Something like sandbox (expo snack) is not possible, because[ I needed to adjust the metro config](https://github.com/bi3ri/rehype-react-native-demo/blob/main/metro.config.js) to support modules, which is not supported for expo snack.

> Are you seeing a typescript error or a runtime error?
> Please share a reproducible example https://stackoverflow.com/help/minimal-reproducible-example

No, there is no error at all. It just ignores the overwriting of the component. (e.g. case 'h2')

> You probably will, both markdown and HTML allow text in places that react-native does not.
> You'll need to restructure text to ensure it is wrapped in a <Text> component.
> 
> Related previous discussions:
> https://github.com/orgs/remarkjs/discussions/1002
> https://github.com/orgs/remarkjs/discussions/1183

Yeah I also found these discussions! They are very helpful, thanks!  :+1: :pray: 

So if I want to create my own processor I would need to develop the following packages:
1) plugin for `hast` tree which transforms `text` into `elements`
2) plugin, to deal with the `<View>` inside a `<Text>` component problem
3) `hast-util-to-react-native` helper package
4) `rehype-react-native` as processor

I also crossed the package [`mdast-normalize-react-native`](https://www.npmjs.com/search?q=remark%20react%20native), which will probably will be very helpful as well.

I am planing to use `remark-rehype` and `mdast-util-to-hast` as blueprint for my packages. 

Any thoughts or suggestions for my package separation? :) 

> It's also worth noting web views also exist I'm React native and can display browser rendered HTML snippets in native content https://github.com/react-native-webview/react-native-webview

My application is using native only components, so a webview is unfortunately not working in my case.

**Reply by wooorm** [reply]
If you have typescript in your repo, or console.logs in your component, you should see what’s up.

If you pass a component for `h2`, you will get certain props, right? And it includes `node`.
That node **does not include `value`**.
If you `console.log(node)`, you will see which fields that object contains.
But this is going in the wrong direction.

These projects have already compiled the AST to the things that are rendered.
With components, you can hook into the final phase.
But now you’re trying to get back into the AST.

The question that remains is: what do you want to make? Do you only want to overwrite `h2`?
Later in your comment here you have two things: a) turn next nodes into elements; b) something about views in texts?
For a), do you really want to wrap every text? For `a *b* c`, should that be `<p><Text>a </Text><em><Text>b</Text></em><Text> c</Text></p>`? Something else?
For b), I don’t understand, what about `View`?

**Reply by bi3ri** [reply]
> If you have typescript in your repo, or console.logs in your component, you should see what’s up.
> 
> If you pass a component for h2, you will get certain props, right? And it includes node.
> That node does not include value.
> If you console.log(node), you will see which fields that object contains.
> But this is going in the wrong direction.

There is no console output for `console.log(node)` inside `h2` at all.  
That's why I asked if it is even possible to pass a component in that case, as it seems that it is ignored.
If node does not include value I would get at least an error, but I don't. 
 
> These projects have already compiled the AST to the things that are rendered.
> With components, you can hook into the final phase.
> But now you’re trying to get back into the AST.

The projects mentioned should only help me how to structure and code my own. That's what i meant with blueprint.  I don't want to integrate them in my project. 

> The question that remains is: what do you want to make? Do you only want to overwrite h2?
> Later in your comment here you have two things: a) turn next nodes into elements; b) something about views in texts?
> For a), do you really want to wrap every text? For a *b* c, should that be <p><Text>a </Text><em><Text>b</Text></em><Text> c</Text></p>? Something else?
> For b), I don’t understand, what about View?

At its core, I want to render Markdown in React Native and am searching for the easiest solution that fits the needs of my project. Now, the question is whether it makes sense to use `rehype-react` and overwrite every component or to write my own processor from scratch. As it seems that overwriting components with other components is not possible, it is probably the latter.

For a) It should be sth like this `<h1>Hello World!<h1>` => `<View><Text style={{fontSize: 30}}>Hello World!</Text></View>`
For b) A View nested inside a Text element throws an error e.g.: `<Text><View> ... </View><Text>`

**Reply by wooorm** [reply]
> There is no console output for console.log(node) inside h2 at all.

In your demo there is no “h2” in the markdown. Components only get used if their corresponding thing is used.

> As it seems that overwriting components with other components is not possible,

It’s quite possible.

> a) 

I specifically wonder what you want to do with emphasis in there

> b)

Do you want every heading and paragraph turned into a separate view? Or is there one view for the whole?

**Reply by bi3ri** [reply]
> In your demo there is no “h2” in the markdown. Components only get used if their corresponding thing is used.

That is right, sorry! I tried many things and this was the last. I did also try the `h2` case e.g. "## Hello World!", though with the same result.

> It’s quite possible.

Oh okay, then there must be a problem somewhere else. Do you have an idea where that could be?

> I specifically wonder what you want to do with emphasis in there

Mhm, isn't `<em>` the tag for emphasis? I dont't see where I am using them.

> Do you want every heading and paragraph turned into a separate view? Or is there one view for the whole?

That's what a different library did, but you are right probably you don't need to wrap the `<Text>` element inside `<View>` element.

**Reply by wooorm** [reply]
> That is right, sorry! I tried many things and this was the last. I did also try the h2 case e.g. `"## Hello World!"`, though with the same result.

> Oh okay, then there must be a problem somewhere else. Do you have an idea where that could be?

Indeed there is another problem!

You made a function `useRemarkSync` that returns the result of `processSync`.
You call `console.log` on what it returns, and see in your console that a particular object is logged.

If you read the readme about `processSync`, or the readme of `rehype-react`, you will find that `processSync` returns a file object. On that file, there is a `result` property. *That* is where the result is.

React does not know about our files. React only knows about the thing at `result`.

```diff
diff --git a/App.js b/App.js
index 16c438f..344c6dc 100644
--- a/App.js
+++ b/App.js
@@ -25,6 +25,7 @@ const components = {
   // this does not work
   h2: (props) => {
     const { node } = props;
+    console.log('called: ', arguments)
     return <Text>{node.value}</Text>;
   },
 };
@@ -37,7 +38,8 @@ export const useRemarkSync = (source = {}) =>
     .processSync(source);
 
 export default function App() {
-  const content = useRemarkSync("# Hello World");
+  const file = useRemarkSync("## Hello World");
+  const content = file.result;
   console.log(content);
 
   return <View style={styles.container}>{content}</View>;
```

> Mhm, isn't `<em>` the tag for emphasis? I dont't see where I am using them.

That is correct! You are not using `<em>`. But I need to know how you want to handle that case. Otherwise I can’t advise you best on what to do.

**Reply by bi3ri** [reply]
Oh, okay, now I understand what the problem was! The transformation occurs when the result is rendered. I had thought it happened before. That's why I wasn't too concerned about the result I was rendering; I assumed that without a transformation, it would throw errors anyway.

```javascript
const components = {

  // this does not work
  h2: (props) => {
    console.log("--- inside h2 ---")
    console.log("props: ", props);
    const { children } = props;
    return <Text>{children}</Text>;
  },
};

export const useRemarkSync = (source = {}) =>
  unified()
    .use(remarkParse)
    .use(remarkToRehype)
    .use(rehypeReact, { ...production, components })
    .processSync(source);

export default function App() {
  const file = useRemarkSync("## Hello World!");
  console.log("file: ", file);
  const content = file.result;
  console.log("content: ", content);

  return <View style={styles.container}>{content}</View>;
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center",
  },
});
```

This produces the following logs:
```bash
 LOG  file:  {"cwd": "/", "data": {}, "history": [], "messages": [], "result": <React.Fragment><h2>Hello World</h2></React.Fragment>, "value": "## Hello World"}
 LOG  content:  <React.Fragment><h2>Hello World</h2></React.Fragment>
 LOG  --- inside h2 ---
 LOG  props:  {"children": "Hello World"}
Android Bundling complete 44ms
 LOG  file:  {"cwd": "/", "data": {}, "history": [], "messages": [], "result": <React.Fragment><h2>Hello World!</h2></React.Fragment>, "value": "## Hello World!"}
 LOG  content:  <React.Fragment><h2>Hello World!</h2></React.Fragment>
 LOG  --- inside h2 ---
 LOG  props:  {"children": "Hello World!"}
```

This is particularly confusing because when I define this component:
```javascript
const components = {
  // this works
  h1: "h4",
};
```

And change the input markdown to "# Hello World!", it generates the following logs:

```bash
 LOG  file:  {"cwd": "/", "data": {}, "history": [], "messages": [], "result": <React.Fragment><h4>rHello World!</h4></React.Fragment>, "value": "# Hello World!"}
 LOG  content:  <React.Fragment><h4>rHello World!</h4></React.Fragment>
 ERROR  Error: Text strings must be rendered within a <Text> component.
```

So in this case the transform happens before rendering. 
Thanks, for the help! :)

> That is correct! You are not using `<em>`. But I need to know how you want to handle that case. Otherwise I can’t advise you best on what to do.

I haven't given it much thought yet, but in other projects, they just use nested <Text> elements with the style set to italic, for example:

```javascript 
<Text> This is normal text... <Text style={{ fontStyle: "italic" }}> this text is italic </Text></Text>
```

By default, there is no style inheritance in React Native, [except for Text subtrees](https://reactnative.dev/docs/text#limited-style-inheritance). So the approach above should also work if you nest other elements inside each other, like an emphasis inside a heading, for example.

**Reply by wooorm** [reply]
Looks like you’re on your way now. Good luck! Some more thoughts if you get stuck:

> The transformation occurs when the result is rendered.

Not quite I think, “the” transformation has happened.
But in React, components are called when they are rendered.
And when changes happen, they get rendered again.
If a subtree doesn’t render, React doesn’t call the components inside it.
You could call rendering “transformation” but I don’t think that’s the right model to think about how React works.

You’re asking a lot about the symptomps, but not about what *actual* input should lead to what *actual* output. What about `em` in a heading? Everything in a `View` or one `View` with everything in it? I think it will help to look at a sample of all that’s possible in markdown and then look at what you’d like to show it as. Emphasis in links? Block quotes? List items? Etc.

> for example:

Can you use the markdown syntax for code to format your code? Now it gets interpreted as markdown.
The text resetting might be problematic when emphasis is used inside strong, or emphasis in a link, etc. (`*a **b** c*`, `[a *b* c](url)`)
